File: optimize-images.js
// optimize-images.js
// Usage: node optimize-images.js
// Requires: npm install sharp
// Scans all HTML files + custom.css for referenced images, then:
//   1. Re-encodes referenced WebP files >100KB at 85% quality (90% for LCP hero)
//   2. Lists PNG files safe to delete (unreferenced with WebP counterparts)
//   3. Flags large unreferenced files

const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const IMAGES_DIR = path.join(__dirname, 'images');
const LCP_IMAGE = 'lawn7.webp';
const SKIP_BELOW_KB = 100;

function getReferencedImages() {
  const referenced = new Set();

  // Scan all .html files in root directory
  const htmlFiles = fs.readdirSync(__dirname).filter(f => f.endsWith('.html'));
  for (const file of htmlFiles) {
    const content = fs.readFileSync(path.join(__dirname, file), 'utf8');
    // Match src="images/xxx", href="images/xxx", url(images/xxx), url("images/xxx")
    for (const m of content.matchAll(/(?:src|href|content|url)\s*[=\(]["']?images\/([^"'\s\)>?#]+)/g)) {
      referenced.add(m[1]);
    }
  }

  // Scan custom.css for background-image references (../images/xxx)
  try {
    const css = fs.readFileSync(path.join(__dirname, 'css', 'custom.css'), 'utf8');
    for (const m of css.matchAll(/url\(['"]?\.\.\/images\/([^'")\s]+)['"]?\)/g)) {
      referenced.add(m[1]);
    }
  } catch (e) {
    console.warn('Could not read css/custom.css:', e.message);
  }

  return referenced;
}

async function compressWebP(referenced) {
  const webpFiles = fs.readdirSync(IMAGES_DIR)
    .filter(f => f.toLowerCase().endsWith('.webp'))
    .sort();

  let totalSaved = 0;
  let compressed = 0;
  let skipped = 0;

  for (const file of webpFiles) {
    if (!referenced.has(file)) {
      continue; // Skip unreferenced files
    }

    const fp = path.join(IMAGES_DIR, file);
    const stat = fs.statSync(fp);

    if (stat.size < SKIP_BELOW_KB * 1024) {
      console.log(`SKIP  (${(stat.size/1024).toFixed(0).padStart(5)}KB) ${file}`);
      skipped++;
      continue;
    }

    const quality = file === LCP_IMAGE ? 90 : 85;
    const tmp = fp + '.tmp.webp';

    try {
      await sharp(fp)
        .webp({ quality, effort: 6 })
        .toFile(tmp);

      const newStat = fs.statSync(tmp);
      const saved = stat.size - newStat.size;
      const pct = ((saved / stat.size) * 100).toFixed(1);

      console.log(`COMPRESS (${(stat.size/1024).toFixed(0).padStart(5)}KB → ${(newStat.size/1024).toFixed(0).padStart(5)}KB, -${pct}%, q=${quality}) ${file}`);

      fs.renameSync(tmp, fp);
      totalSaved += saved;
      compressed++;
    } catch (e) {
      console.error(`ERROR: ${file}: ${e.message}`);
      if (fs.existsSync(tmp)) fs.unlinkSync(tmp);
    }
  }

  console.log(`\nCompressed ${compressed} files. Skipped ${skipped} (under ${SKIP_BELOW_KB}KB).`);
  console.log(`Total savings: ${(totalSaved / 1024 / 1024).toFixed(2)}MB`);
}

function listDeletablePNGs(referenced) {
  console.log('\n=== PNG FILES SAFE TO DELETE ===');
  const pngFiles = fs.readdirSync(IMAGES_DIR).filter(f => f.toLowerCase().endsWith('.png')).sort();
  let totalSize = 0;

  for (const file of pngFiles) {
    const webp = file.replace(/\.png$/i, '.webp');
    const webpPath = path.join(IMAGES_DIR, webp);
    const pngReferenced = referenced.has(file);
    const webpExists = fs.existsSync(webpPath);
    const webpReferenced = referenced.has(webp);

    if (!pngReferenced) {
      const size = fs.statSync(path.join(IMAGES_DIR, file)).size;
      totalSize += size;
      const reason = webpExists && webpReferenced
        ? '(WebP counterpart referenced)'
        : webpExists
          ? '(WebP exists but also unreferenced)'
          : '(no WebP counterpart, unreferenced)';
      console.log(`DELETE: images/${file} (${(size/1024/1024).toFixed(1)}MB) ${reason}`);
    }
  }

  console.log(`\nTotal PNG deletion savings: ${(totalSize / 1024 / 1024).toFixed(1)}MB`);
}

function listJunkFiles(referenced) {
  console.log('\n=== OTHER JUNK/LARGE UNREFERENCED FILES ===');
  const allFiles = fs.readdirSync(IMAGES_DIR);

  for (const file of allFiles) {
    // Flag large unreferenced non-image files
    const ext = path.extname(file).toLowerCase();
    if (['.psd', '.ini', '.mp4'].includes(ext) || file.includes(' - Copy') || file.includes(' copy')) {
      const size = fs.statSync(path.join(IMAGES_DIR, file)).size;
      console.log(`JUNK: images/${file} (${(size/1024).toFixed(0)}KB)`);
    }
    // Flag large unreferenced SVGs
    if (ext === '.svg' && !referenced.has(file)) {
      const size = fs.statSync(path.join(IMAGES_DIR, file)).size;
      if (size > 50 * 1024) {
        console.log(`LARGE UNREFERENCED SVG: images/${file} (${(size/1024).toFixed(0)}KB)`);
      }
    }
  }
}

(async () => {
  console.log('=== No-BS Yardwork Image Optimizer ===\n');
  console.log('Scanning HTML files and CSS for referenced images...');
  const referenced = getReferencedImages();
  console.log(`Found ${referenced.size} unique referenced images.\n`);

  console.log('=== COMPRESSING REFERENCED WebP FILES ===');
  await compressWebP(referenced);

  listDeletablePNGs(referenced);
  listJunkFiles(referenced);

  console.log('\n=== DONE ===');
})();

File: optimize-images.py
#!/usr/bin/env python3
"""
optimize-images.py — No-BS Yardwork Image Optimizer
Usage: python optimize-images.py
Requires: pip install Pillow

1. Scans HTML files + custom.css for referenced images
2. Re-encodes referenced WebP files >100KB (85% quality, 90% for LCP hero)
3. Lists PNG files safe to delete (no HTML references, WebP counterpart exists)
4. Lists junk files (copies, PSD, ini, etc.)
"""

import os
import re
import sys
import shutil
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    print("ERROR: Pillow not installed. Run: pip install Pillow")
    sys.exit(1)

ROOT = Path(__file__).parent
IMAGES_DIR = ROOT / "images"
LCP_IMAGE = "lawn7.webp"
SKIP_BELOW_KB = 100


def get_referenced_images():
    referenced = set()

    # Scan all HTML files in root
    for html_file in ROOT.glob("*.html"):
        content = html_file.read_text(encoding="utf-8", errors="ignore")
        for m in re.finditer(r'(?:src|href|content|url)\s*[=(]["\']?images/([^"\'\s)>?#]+)', content):
            referenced.add(m.group(1))

    # Scan custom.css for background-image references
    css_path = ROOT / "css" / "custom.css"
    if css_path.exists():
        css = css_path.read_text(encoding="utf-8", errors="ignore")
        for m in re.finditer(r'url\([\'"]?\.\.\/images\/([^\'")\s]+)[\'"]?\)', css):
            referenced.add(m.group(1))

    return referenced


def compress_webp(referenced):
    print("=== COMPRESSING REFERENCED WebP FILES ===")
    total_saved = 0
    compressed = 0
    skipped = 0

    webp_files = sorted(IMAGES_DIR.glob("*.webp"))

    for fp in webp_files:
        if fp.name not in referenced:
            continue

        size = fp.stat().st_size
        if size < SKIP_BELOW_KB * 1024:
            print(f"SKIP  ({size//1024:>5}KB) {fp.name}")
            skipped += 1
            continue

        quality = 90 if fp.name == LCP_IMAGE else 85
        tmp = fp.with_suffix(".tmp.webp")

        try:
            with Image.open(fp) as img:
                # Convert to RGB/RGBA to strip EXIF metadata (smaller file, no charset issues)
                if img.mode in ("RGBA", "LA", "P"):
                    img = img.convert("RGBA")
                else:
                    img = img.convert("RGB")
                img.save(tmp, "WEBP", quality=quality, method=6)

            new_size = tmp.stat().st_size
            saved = size - new_size
            pct = (saved / size * 100) if size > 0 else 0

            if new_size >= size:
                print(f"SKIP  (already optimal: {size//1024}KB, re-encoded={new_size//1024}KB) {fp.name}")
                tmp.unlink()
                skipped += 1
                continue

            print(f"COMPRESS ({size//1024:>5}KB → {new_size//1024:>5}KB, -{pct:.1f}%, q={quality}) {fp.name}")

            shutil.move(str(tmp), str(fp))
            total_saved += saved
            compressed += 1

        except Exception as e:
            print(f"ERROR: {fp.name}: {e}")
            if tmp.exists():
                tmp.unlink()

    print(f"\nCompressed {compressed} files. Skipped {skipped} (under {SKIP_BELOW_KB}KB).")
    print(f"Total savings: {total_saved / 1024 / 1024:.2f}MB")


def list_deletable_pngs(referenced):
    print("\n=== PNG FILES SAFE TO DELETE ===")
    total_size = 0

    for fp in sorted(IMAGES_DIR.glob("*.png")):
        if fp.name in referenced:
            continue  # PNG is directly referenced — don't delete

        webp_name = fp.stem + ".webp"
        webp_path = IMAGES_DIR / webp_name
        webp_exists = webp_path.exists()
        webp_referenced = webp_name in referenced

        size = fp.stat().st_size
        total_size += size

        if webp_exists and webp_referenced:
            reason = "(WebP counterpart referenced in HTML)"
        elif webp_exists:
            reason = "(WebP exists but also unreferenced)"
        else:
            reason = "(no WebP counterpart, fully unreferenced)"

        print(f"DELETE: images/{fp.name} ({size/1024/1024:.1f}MB) {reason}")

    print(f"\nTotal PNG deletion savings: {total_size/1024/1024:.1f}MB")


def list_junk_files(referenced):
    print("\n=== JUNK / LARGE UNREFERENCED FILES ===")
    for fp in sorted(IMAGES_DIR.iterdir()):
        if not fp.is_file():
            continue
        ext = fp.suffix.lower()
        name = fp.name
        size = fp.stat().st_size

        is_junk = (
            ext in (".psd", ".ini", ".mp4")
            or " - Copy" in name
            or " copy" in name.lower()
            or name == "desktop.ini"
        )
        is_large_unreferenced_svg = (
            ext == ".svg"
            and name not in referenced
            and size > 50 * 1024
        )

        if is_junk:
            print(f"JUNK:   images/{name} ({size//1024}KB)")
        elif is_large_unreferenced_svg:
            print(f"LARGE UNREFERENCED SVG: images/{name} ({size//1024}KB)")


def main():
    print("=== No-BS Yardwork Image Optimizer ===\n")
    print("Scanning HTML files and CSS for referenced images...")
    referenced = get_referenced_images()
    print(f"Found {len(referenced)} unique referenced images.\n")

    compress_webp(referenced)
    list_deletable_pngs(referenced)
    list_junk_files(referenced)

    print("\n=== DONE ===")


if __name__ == "__main__":
    main()
